mlir/lib/Dialect/Rock/Transforms/TransformToMemref.cpp

```diff
@@ -0,0 +1,132 @@
+//===-----------------------------------------------------===//
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Rock/IR/Rock.h"
+#include "mlir/Dialect/Rock/IR/TransformMapBuilder.h"
+#include "mlir/Dialect/Rock/Passes.h"
+#include "mlir/Dialect/Rock/Tuning/GeneralGemmBlockStructure.h"
+#include "mlir/Dialect/Rock/utility/builderUtils.h"
+#include "mlir/Dialect/Rock/utility/loweringUtils.h"
+#include "mlir/Dialect/Rock/utility/transformMapUtils.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Rock/IR/MfmaInsnGroup.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Debug.h"
+
+namespace mlir {
+namespace rock {
+#define GEN_PASS_DEF_ROCKTRANSFORMTOMEMREFPASS
+#include "mlir/Dialect/Rock/Passes.h.inc"
+} // namespace rock
+} // namespace mlir
+
+#define DEBUG_TYPE "rock-transform-to-memref"
+
+using namespace mlir;
+using namespace mlir::arith;
+using namespace mlir::rock;
+
+namespace {
+struct RockTransformToMemrefPass
+    : public rock::impl::RockTransformToMemrefPassBase<
+          RockTransformToMemrefPass> {
+  void runOnOperation() override;
+};
+
+//===----------------------------------------------------------------------===//
+// TransformOp conversion to MemRef
+//===----------------------------------------------------------------------===//
+struct TransformRewritePattern : public OpRewritePattern<TransformOp> {
+  using OpRewritePattern<TransformOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(TransformOp op,
+                                PatternRewriter &b) const override {
+    Location loc = op.getLoc();
+    auto srcType = op.getOperand().getType().cast<ShapedType>();
+    auto resType = op.getResult().getType().cast<ShapedType>();
+
+    bool expanded = resType.getShape().size() > srcType.getShape().size();
+    SmallVector<ReassociationIndices> merges(
+        expanded ? srcType.getShape().size() : resType.getShape().size());
+
+    // only converts simple expand/collapse form
+    for (auto tattr : op.getTransform().getOps()) {
+      auto inDims = expanded ? tattr.getLowerDims() : tattr.getUpperDims();
+      auto outDims = expanded ? tattr.getUpperDims() : tattr.getLowerDims();
+      switch (tattr.getType()) {
+      case rock::TransformType::PassThrough:
+        for (auto pair : llvm::zip(inDims, outDims)) {
+          auto inDim = std::get<0>(pair);
+          assert(merges[inDim].empty());
+          merges[inDim].push_back(std::get<1>(pair));
+        }
+        break;
+      case rock::TransformType::Pad:
+      case rock::TransformType::Slice:
+      case rock::TransformType::Embed:
+      case rock::TransformType::Broadcast: // Unsupported
+      case rock::TransformType::AddDim:
+      case rock::TransformType::ConstDim:
```

Comment:
AddDim and ConstDim aren't failure cases - they come up when going to/from a scalar memref like `memref<f32>`. That should be handled here.

Reply:
Agreed, but we only see this case for collapse_shape in the init kernel.

Reply:
then I will need a test..